sync_server_architecture.adoc

```diff
@@ -0,0 +1,72 @@
+[[sync-server-architecture]]
+= Sync Server Architecture
+
+[[overview]]
+== Overview
+
+* The sync server provides default data access implementations using backend
+data store which the developer can override to store sync record data in
+data stores other than the default store.
+* Collision handlers can also be overridden on a per-dataset level.
+** List Collision, Remove Collision, Handle Collision.
+* The hashing algorithm used for both global and individual sync records can be
+set independently per dataset.
+* The sync server stores data state in backend data store, allowing for
+the sync server to be horizontally scalable.
+** Uses queues in backend data store to manage workers syncing while
+horizontally scaled.
+
+[[architecture]]
+== Architecture
+
+The inner-workings of the sync server architecture consist of HTTP handlers,
+queues, processors and the sync scheduler.
+Each of these components persist any data they need in a MongoDB data store.
+
+.Inner-workings of the sync server architecture
+image::figures/sync_architecture.png[]
+
+=== HTTP Handlers
+These handlers are responsible handling the sync requests from sync clients.
+
+==== Sync HTTP Handler
+Creates or updates the dataset client to be synced and pushes pending records
+and acknowledgements onto their respective queues, so that they will be
+processed later.
+
+Request and response interceptors can be set independently for each dataset.
+
+==== Sync Records HTTP Handler
+Compare up-to-date data with clients state. After getting the delta, iterate
+through all those records, and if any of them pending queue already, remove them
```

Comment:
it also needs to check if there are any updates that are processed but not yet synced.